apollo-federation/src/merge.rs

```diff
@@ -600,23 +647,128 @@ impl Merger {
     }
 
     fn merge_scalar_type(
-        &self,
+        &mut self,
         types: &mut IndexMap<Name, ExtendedType>,
+        directive_names: &DirectiveNames,
         subgraph_name: Name,
         scalar_name: NamedType,
         ty: &Node<ScalarType>,
     ) {
         let existing_type = types
             .entry(scalar_name.clone())
-            .or_insert(copy_scalar_type(scalar_name, ty));
+            .or_insert(copy_scalar_type(scalar_name.clone(), ty));
+
         if let ExtendedType::Scalar(s) = existing_type {
             let join_type_directives =
                 join_type_applied_directive(subgraph_name.clone(), iter::empty(), false);
             s.make_mut().directives.extend(join_type_directives);
+            self.add_inaccessible(
+                directive_names,
+                &mut s.make_mut().directives,
+                &ty.directives,
+            );
         } else {
             // conflict?
         }
     }
+
+    fn add_inaccessible(
+        &mut self,
+        directive_names: &DirectiveNames,
+        new_directives: &mut apollo_compiler::schema::DirectiveList,
+        original_directives: &apollo_compiler::schema::DirectiveList,
+    ) {
+        if original_directives
+            .iter()
+            .any(|d| d.name == directive_names.inaccessible)
+        {
+            self.needs_inaccessible = true;
+
+            new_directives.push(Component::new(Directive {
+                name: INACCESSIBLE_DIRECTIVE_NAME_IN_SPEC,
+                arguments: vec![],
+            }));
+        }
+    }
+
+    fn add_inaccessible_ast(
+        &mut self,
+        directive_names: &DirectiveNames,
+        new_directives: &mut DirectiveList,
+        original_directives: &DirectiveList,
+    ) {
+        if original_directives
+            .iter()
+            .any(|d| d.name == directive_names.inaccessible)
+        {
+            self.needs_inaccessible = true;
+
+            new_directives.push(
+                Directive {
+                    name: INACCESSIBLE_DIRECTIVE_NAME_IN_SPEC,
+                    arguments: vec![],
+                }
+                .into(),
+            );
+        }
+    }
```

Comment:
Small nit: this could be genericized. Works as is if you want to keep it instead!